diffrax/solver/additive_srk.py

```diff
@@ -0,0 +1,284 @@
+    a: list[np.ndarray]
+
+    # coefficients for W and H (of shape (len(c)+1,)
+    cW: np.ndarray
+    cH: Optional[np.ndarray]
+    bW: Scalar
+    bH: Optional[Scalar]
+
+    def __post_init__(self):
+        assert self.c.ndim == 1
+        for a_i in self.a:
+            assert a_i.ndim == 1
+        assert self.b_sol.ndim == 1
+        assert (self.b_error is None) or self.b_error.ndim == 1
+        assert self.c.shape[0] == len(self.a)
+        assert all(i + 1 == a_i.shape[0] for i, a_i in enumerate(self.a))
+        assert (self.b_error is None) or self.b_error.shape[0] == self.b_sol.shape[0]
+        assert self.c.shape[0] + 1 == self.b_sol.shape[0]
+        assert self.cW.shape[0] == self.b_sol.shape[0]
+        assert self.cH.shape[0] == self.b_sol.shape[0]
+        for i, (a_i, c_i) in enumerate(zip(self.a, self.c)):
+            assert np.allclose(sum(a_i), c_i)
+        assert np.allclose(sum(self.b_sol), 1.0)
+
+        # TODO: add checks for whether the method is FSAL
+
+
+StochasticButcherTableau.__init__.__doc__ = """**Arguments:**
+
+Let `k` denote the number of stages of the solver.
+
+- `a`: the lower triangle (without the diagonal) of the Butcher tableau. Should
+    be a tuple of NumPy arrays, corresponding to the rows of this lower triangle. The
+    first array represents the should be of shape `(1,)`. Each subsequent array should
+    be of shape `(2,)`, `(3,)` etc. The final array should have shape `(k - 1,)`.
+- `b_sol`: the linear combination of stages to take to produce the output at each step.
+    Should be a NumPy array of shape `(k,)`.
+- `b_error`: the linear combination of stages to take to produce the error estimate at
+    each step. Should be a NumPy array of shape `(k,)`. Note that this is *not*
+    differenced against `b_sol` prior to evaluation. (i.e. `b_error` gives the linear
+    combination for producing the error estimate directly, not for producing some
+    alternate solution that is compared against the main solution).
+- `c`: the time increments used in the Butcher tableau.
+    Should be a NumPy array of shape `(k-1,)`, as the first stage has time increment 0
+- `cW`: The coefficients in front of the Brownian increment at each stage.
+    Should be a NumPy array of shape `(k,)`.
+- `cH`: The coefficients in front of the space-time Lévy area at each stage.
+    Should be a NumPy array of shape `(k,)`.
+- `bW`: Coefficient for the Brownian increment when computing the
+    output $y_{n+1}$. Should be a `Scalar`.
+- `bH`: Coefficient for the space-time Lévy area when computing the
+    output $y_{n+1}$. Should be a `Scalar`.
+"""
+
+
+class AbstractAdditiveSRK(AbstractStratonovichSolver):
+    r"""Additive-Noise Stochastic Runge-Kutta method.
+
+    The second term in the MultiTerm must be a `ControlTerm` with
+    `control=VirtualBrownianTree(levy_area="space-time")`, since this method
+    makes use of space-time Lévy area.
+
+    Given the SDE
+    $dX_t = f(t, X_t) dt + σ dW_t$
```

Comment:
Just for my own education: is there actually any scenario in which time-dependent and time-independent additive noise need to be treated differently?

As for a general SRK -- if this is landing later, perhaps we should remove `AbstractAdditiveSRK` from the public interface now, on the assumption it will be removed later?

Reply:
On your first paragraph: no strong feelings; up to you.

But yes, definitely do remove it from the docs / public API if it's not going to be long for this world!

---

Do you think it would it be worth splitting this PR into two pieces? Keep just the BM changes here, and then do the solvers separately? The BM changes are now very nearly ready, and I'm thinking it might not be worth ironing out the details of the solvers if they're going to be changing soon anyway. (The bottleneck here is clearly my review time, so anything that minimises that is good :) )

Reply:
Yes, I can try to split it, that's probably a good idea. What's the best way to accomplish this? Should I just start from a fresh branch (up to date with just main Diffrax) and then add in all the brownian-related files, and not the solvers? There are some files (e.g. term.py) which have changes relating to both one and the other, and I think it would be a mess trying to separate all of that out. What do you suggest (mostly in terms of the git procedure + useful commands)?

Reply:
> Should I just start from a fresh branch (up to date with just main Diffrax) and then add in all the brownian-related files, and not the solvers? 

Yup, exactly.

For `term.py`, let's leave that out for now as it isn't reviewed yet.

In terms of git commands: the simplest way is probably to copy-paste the browian files somewhere else, then create a new branch, then copy them back!